File: src/model/color.js
// Updated 2026-07-24 for JavaScript ES6 refactor.
function isFiniteNumber(value) {
  return typeof value === "number" && isFinite(value);
}

function validateChannel(name, value) {
  if (!isFiniteNumber(value)) {
    throw new TypeError(name + " must be a finite number");
  }

  if (value < 0.0 || value > 1.0) {
    throw new RangeError(name + " must lie between 0.0 and 1.0");
  }
}

class Color {
  constructor(r, g, b, a) {
    this.r = r;
    this.g = g;
    this.b = b;
    this.a = a;

    this.validate();
  }

  validate() {
    validateChannel("r", this.r);
    validateChannel("g", this.g);
    validateChannel("b", this.b);
    validateChannel("a", this.a);
  }

  toArray() {
    return [this.r, this.g, this.b, this.a];
  }

  copy() {
    return new Color(this.r, this.g, this.b, this.a);
  }
}

module.exports = {
  Color
};

File: pySauce/color.py
from dataclasses import dataclass

@dataclass
class Color:
    r: float
    g: float
    b: float
    a: float

    def __post_init__(self):

        self.validate()

    def validate(self):

        for name, value in (

            ("r", self.r),

            ("g", self.g),

            ("b", self.b),

            ("a", self.a),

        ):

            if not isinstance(value, (int, float)):

                raise TypeError(f"{name} must be a number")

            if not (0.0 <= value <= 1.0):

                raise ValueError(f"{name} must lie between 0.0 and 1.0")

    def to_tuple(self):
        return (self.r, self.g, self.b, self.a)
    